Content.Server/Anomaly/AnomalySystem.Behaviour.cs

```diff
@@ -0,0 +1,104 @@
+using System.Linq;
+using Content.Shared.Anomaly.Components;
+using Content.Shared.Anomaly.Prototypes;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization.Manager;
+
+namespace Content.Server.Anomaly;
+public sealed partial class AnomalySystem
+{
+    [Dependency] private readonly IComponentFactory _componentFactory = default!;
+    [Dependency] private readonly ISerializationManager _serialization = default!;
+
+    private List<AnomalyBehaviourPrototype> _behaviourList = new();
+    private void InitializeBehaviour()
+    {
+        SubscribeLocalEvent<PrototypesReloadedEventArgs>(OnPrototypesReloaded);
+
+        //Cache all behaviors into a list at the beginning of the round
+        _behaviourList.AddRange(_prototype.EnumeratePrototypes<AnomalyBehaviourPrototype>());
+    }
+
+    private void OnPrototypesReloaded(PrototypesReloadedEventArgs ev)
+    {
+        _behaviourList.Clear();
+        _behaviourList.AddRange(_prototype.EnumeratePrototypes<AnomalyBehaviourPrototype>());
+    }
+
+    private string GetRandomBehaviour()
+    {
+        var totalWeight = _behaviourList.Sum(x => x.Weight);
+        var randomValue = _random.NextFloat(totalWeight);
+
+        foreach (var b in _behaviourList)
+        {
+            if (randomValue < b.Weight)
+            {
+                return b.ID;
+            }
+
+            randomValue -= b.Weight;
+        }
+        // Shouldn't happen
+        throw new InvalidOperationException($"Invalid weighted variantize anomaly behaviour pick!");
+    }
+
+    private void SetBehaviour(Entity<AnomalyComponent> anomaly, ProtoId<AnomalyBehaviourPrototype> behaviourProto)
+    {
+        if (anomaly.Comp.CurrentBehaviour == behaviourProto)
+            return;
+
+        if (anomaly.Comp.CurrentBehaviour != null)
+            RemoveBehaviour(anomaly);
+
+        anomaly.Comp.CurrentBehaviour = behaviourProto;
+
+        var behaviour = _prototype.Index(behaviourProto);
+
+        foreach (var (name, entry) in behaviour.Components)
+        {
+            var reg = _componentFactory.GetRegistration(name);
+
+            if (EntityManager.HasComponent(anomaly, reg.Type))
+            {
+                EntityManager.RemoveComponent(anomaly, reg.Type);
+            }
+
+            var comp = (Component) _componentFactory.GetComponent(reg);
+            comp.Owner = anomaly;
+
+            var temp = (object) comp;
+            _serialization.CopyTo(entry.Component, ref temp);
+            EntityManager.RemoveComponent(anomaly, temp!.GetType());
+            EntityManager.AddComponent(anomaly, (Component) temp!);
+        }
+    }
```

Comment:
You remove the component then remove it again anyway, also don't do null suppression unless you're sure of what you're doing.